### ISSUE TITLE
Distinguish corruption and partial flush

### ISSUE THREAD

**Issue by MiguelPires** [camunda-cloud/zeebe#6744]
**Description**

Currently, a partial write caused by a crash during a flush is detected as log corruption. The issue is that even though the partial write can be safely discarded, a partially flushed write it's currently indistinguishable from an actual log corruption. This is the cause of this issue #6709 and the short-term solution to unblock us is to ignore it during start-up. Since that means we won't detect corruption, it would be good to have a better long-term solution.

Section 3.3.3 of https://www.usenix.org/system/files/conference/fast18/fast18-alagappan.pdf suggests an additional "persist" write and flush. I think this is equivalent to writing the written index in the config store. However, this would require another write and flush before the record can be considered persisted. The paper then suggests writing both and only flushing once after to reduce the performance impact. This means that if the second write can be verified, the checksum failed because of a partial write. If the second write isn't present then it's either because of a partially flushed write or because both were corrupted. These two situations can only be distinguished if we can read valid entries after them (which would tell us the previous records were corrupted). Corruption in the last record and in its "persist" record, and a partial write can't be distinguished.

**Comment by MiguelPires** [comment]
I'm going to implement the simplest solution, evaluate the performance impact of that and then iterate on that if necessary. If the performance hit is not significant then we avoid more complicated and time-consuming solutions.
The approach aims to distinguish between log corruption and a partial write using an index that will be persisted in the meta store. This index is the index of the last flushed entry in the log. If a checksum mismatch is detected, we can use this to tell if we expected the log to have more entries (log corruption) or if the valid last entry in the log matches the index in the metastore (partial write). If an entry was partially flushed then we know that it wasn't committed and we can mark it as ignored. This requires the last flushed index to be updated whenever the we flush, so after handling an AppendRequest on the followers and after updating the commit index on the leader. I'll open another issue for this since it's not actually solving the problem, just preparing the solution.
